### PR TITLE
Add caret annotations

### LINKED ISSUE #367: Implement Caret Listener and Caret Annotations
Currently, Saros/I is still missing caret annotations (highlighting the caret position of other users). To realize this feature, a caret listener has to be added and the annotation feature has to be implemented.

There is already a placeholder class for a caret listener: [`StoppableCaretListener`](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/eventhandler/editor/caret/StoppableCaretListener.java). This class however implements the wrong interface. The correct interface to use is `com.intellij.openapi.editor.event.CaretListener`. `caretPositionChanged` is most likely the only needed method.

**Update:** This will most likely have to replace the current selection listener as the selection listener does not seem differentiate between carets (needs more testing when working on this). Instead, the caret listener should differentiate between the carets in the event handling (e.g. hold a reference to the added "Saros carets" or check if it is primary cared through `CaretModel.getPrimaryCaret()`). This can replace the selection listener as all selections also automatically include a caret location change. The selection range of the caret can then also be read and manipulated through the `Caret` object.

**Update 2:** Carets can't be used to display remote carets as they always remain controlled by the user, i.e. using the arrows will move *all* carets in the current editor (not only the primary caret). Instead, a zero-width boxed highlighting can be used to simulate the appearance of a caret.

To figure out how the caret information is transmitted to other participants, look at the Saros/E implementation.

#### TODOs
- [x] Implement a caret listener (and possible merge it with `LocalTextSelectionChangeHandler`)
  - [x] Rename `StoppableCaretListener` to match the new naming scheme (e.g. `LocalCaretChangeHandler`)
  - [x] Implement `com.intellij.openapi.editor.event.CaretListener.caretPositionChanged`
- [x] Implement caret annotations
  - [x] If possible/plausible, use the existing annotation logic ([`AbstractEditorAnnotation`](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AbstractEditorAnnotation.java), [`AnnotationStore`](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AnnotationStore.java), [`AnnotationManager`](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AnnotationManager.java))

### PR DESCRIPTION
**Pull request by tobous**
#### [INTERNAL][I] Allow zero-width selection annotations

Allows the addition of zero-width selection annotations.

Adjusts the range for zero-width selection annotations by including the
previous character (if the start is not 0). This was done as a
hack to use selection annotations for caret annotations. This should,
however, be reworked at some point in the future to introduce a handling
that is more resistant to remote edits. With the current handling, the
caret annotation can be removed if its text is removed. This makes sense
for selections but not for carets (as other users can't "delete" your
local caret).

Adjusts the tests to correctly cover this case.

Part of fix for #367.

#### [INTERNAL][I] Add caret position change handler

Adds LocalCaretPositionChangeHandler to create text selection activities
for caret moves.

The handler is disabled internally for files that have an active
selection. This is necessary as the caret listener methods can not be
used to obtain timely information about the current selection. As a
result, we still need to handle such changes in a separate selection
handler.

Removes the old stub StoppableCaretListener.

Resolves #367.

#### [INTERNAL][I] Share caret movement cause by text edits

Adjusts EditorManager.generateTextEdit(...) to also send local selection
of the editor after the edit. This is necessary as the caret listener
that is usually used for this purpose explicitly does not cover caret
moves induced by document changes.

This handling introduces a barely noticeable, weird flickering in the
annotations when a remote user starts the process of renaming something
through the refactoring function. This seems to be unavoidable and is
caused by the internal handling of such refactorings. The name to change
is removed and then re-added to the selected text section. During this
process, the caret position is momentarily set to the start of the file
(position 0). This caret position/selection is then immediately replaced
by a selection of the name to refactor. This is all hidden from the
local instance but still reported to the listeners. As a result, this
behavior is visible in the remote editors. (The removal and re-addition
of the name was already visible beforehand.)

The logic also covers selections and not only caret positions. Even
though the new selection state after a local change by the user will be
of only a caret without an active selection, this can not always be
guaranteed. Document changes could also be cause by refactorings
(started in a different location or a different file). Therefore, the
editor might still have a valid selection after the document
modification that does not have to match the end of the modification. To
also cover these cases, the actual selection state of the editor is
used.

Adds a note pointing to EditorManager.generateTextEdit(...) to the
javadoc of LocalCaretPositionChangeHandler.

Covers last case to resolve #367.

### CODE REVIEW

**Comment by tobous** [comment]
This PR is still missing a further listener dealing with caret movements caused by text edits (as such caret movements are explicitly not covered by the caret listener). Will add it tomorrow (or create another PR for it if it gets too big.)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/eventhandler/editor/caret/LocalCaretPositionChangeHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5475726&bid=17944604)

File: intellij/test/junit/saros/intellij/editor/annotations/AnnotationManagerTest.java [hunk 1]
```diff
@@ -341,6 +341,7 @@ public void testDoNotReplaceSelectionAnnotation() {
    * annotation being added. If an old selection for the file is present, it should still be
    * removed.
    */
+  // TODO adjust test as behavior has changed
```
Comment (tobous) [inline intellij/test/junit/saros/intellij/editor/annotations/AnnotationManagerTest.java:344]: Remove TODO.

**Comment by tobous** [comment]
Rebased onto current base branch without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/eventhandler/editor/caret/LocalCaretPositionChangeHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5475726&bid=17944604)

**Comment by tobous** [comment]
Rebased onto current base branch without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/eventhandler/editor/caret/LocalCaretPositionChangeHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5475726&bid=17944604)

**Comment by tobous** [comment]
Fixed minor issues with adjusted test case.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/eventhandler/editor/caret/LocalCaretPositionChangeHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5475726&bid=17944604)

**Comment by tobous** [comment]
Adjusted for changes in reaction to `isBackwardsSelection` being part of `TextSelection`.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/eventhandler/editor/caret/LocalCaretPositionChangeHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5475726&bid=17944604)

**Comment by tobous** [comment]
Rebased onto current master without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/eventhandler/editor/caret/LocalCaretPositionChangeHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5475726&bid=17944604)

**Comment by tobous** [comment]
Fixed minor code style/doc issues.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/eventhandler/editor/caret/LocalCaretPositionChangeHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5475726&bid=17944604)

**Pull request merged**